Source/NSTreeController.m

```diff
@@ -127,123 +169,258 @@ - (BOOL) selectsInsertedObjects
+  return obj;
 }
 
-- (NSIndexPath*) selectionIndexPath
+- (NSArray *) selectedObjects
 {
-  // FIXME
-  return nil;
+  NSMutableArray *selectedObjects = [NSMutableArray array];
+
+  FOR_IN(NSIndexPath*, path, _selection_index_paths)
+    {
+      id obj = [self _objectAtIndexPath: path];
+      [selectedObjects addObject: obj];
+    }
+  END_FOR_IN(_selection_index_paths);
+
+  return selectedObjects;
 }
 
-- (NSArray*) selectionIndexPaths
+- (NSIndexPath *) selectionIndexPath
 {
-  // FIXME
-  return nil;
+  return [_selection_index_paths objectAtIndex: 0];
+}
+
+- (NSArray *) selectionIndexPaths
+{
+  return _selection_index_paths;
 }
 
-- (NSArray*) sortDescriptors
+- (NSArray *) sortDescriptors
 {
   return _sortDescriptors;
 }
 
-- (NSString*) childrenKeyPath
+- (NSString *) childrenKeyPath
 {
   return _childrenKeyPath;
 }
 
-- (NSString*) countKeyPath
+- (NSString *) countKeyPath
 {
   return _countKeyPath;;
 }
 
-- (NSString*) leafKeyPath
+- (NSString *) leafKeyPath
 {
   return _leafKeyPath;
 }
 
-- (void) addChild: (id)sender
+- (IBAction) add: (id)sender
 {
-  // FIXME
-}
+  if ([self canAddChild]
+      && [self countKeyPath] == nil)
+    {
+      id newObject = [self newObject];
 
-- (void) add: (id)sender
-{
-  // FIXME
-  [super add: sender];
+      if (newObject != nil)
+	{
+	  NSMutableArray *newContent = [NSMutableArray arrayWithArray: [self content]];
+	  GSControllerTreeProxy *node = [[GSControllerTreeProxy alloc]
+					  initWithContent: newObject
+					   withController: self];
+
+	  [newContent addObject: node];
+
+	  [self setContent: newContent];
+	  RELEASE(newObject);
+	}
+    }
 }
 
-- (void) insertChild: (id)sender
+- (IBAction) addChild: (id)sender
 {
-  // FIXME
+  NSIndexPath *p = [self selectionIndexPath];
+  id newObject = [self newObject];
+
+  if (p != nil)
+    {
+      [self insertObject: newObject atArrangedObjectIndexPath: p];
+    }
 }
 
-- (void) insertObject: (id)object atArrangedObjectIndexPath: (NSIndexPath*)indexPath
+- (IBAction) remove: (id)sender
 {
-  // FIXME
+  if ([self canRemove]
+      && [self countKeyPath] == nil)
+    {
+      [self removeObject: [self content]];
+    }
 }
 
-- (void) insertObjects: (NSArray*)objects atArrangedObjectIndexPaths: (NSArray*)indexPaths
+- (IBAction) insertChild: (id)sender
 {
-  // FIXME
+  [self addChild: sender];
 }
 
-- (void) insert: (id)sender
+- (void) insertObject: (id)object atArrangedObjectIndexPath: (NSIndexPath *)indexPath
 {
-  // FIXME
+  if ([self canAddChild]
+      && [self countKeyPath] == nil)
+    {
+      NSUInteger length = [indexPath length] - 1;
+      NSUInteger pos = 0;
+      NSMutableArray *children = [_arranged_objects mutableChildNodes];
+      NSUInteger lastIndex = 0;
+
+      for (pos = 0; pos < length; pos++)
+	{
+	  NSUInteger i = [indexPath indexAtPosition: pos];
+	  id node = [children objectAtIndex: i];
+
+	  children = [node valueForKeyPath: _childrenKeyPath];
+	}
+
+      lastIndex = [indexPath indexAtPosition: length];
```

Comment:
This code here looks the same as the code inside the loop. If this is correct you could just run the loop one more iteration. Which could be achieved by removing the `- 1` when computing the `length`.

Reply:
Fixed